File: python/railjson_generator/railjson_generator/schema/infra/operational_point.py
from dataclasses import dataclass
from typing import Annotated

from osrd_schemas import models
from pydantic import StringConstraints

NonBlankStr = Annotated[str, StringConstraints(min_length=1)]


@dataclass
class OperationalPoint:
    label: NonBlankStr
    id: str
    parts: list
    weight: int | None
    uic: int | None
    plc: NonBlankStr | None
    country_code: NonBlankStr
    main_code: NonBlankStr
    secondary_code: NonBlankStr | None
    secondary_name: NonBlankStr | None
    is_passenger_station: bool

    def __init__(
        self,
        label: str,
        id: str | None = None,
        main_code: NonBlankStr | None = None,
        uic: int | None = 8700,
        plc: NonBlankStr | None = None,
        weight: int | None = None,
        secondary_code: NonBlankStr | None = "BV",
        secondary_name: NonBlankStr | None = "0",
        country_code: NonBlankStr = "FR",
        is_passenger_station: bool = True,
    ):
        self.label = label
        self.main_code = main_code or label[:3].upper()
        self.parts = []
        self.uic = uic
        self.weight = weight
        self.id = id or label
        self.secondary_code = secondary_code
        self.secondary_name = secondary_name
        self.plc = plc
        self.country_code = country_code
        self.is_passenger_station = is_passenger_station

    def add_part(self, track, offset, local_track_name):
        op_part = OperationalPointPart(self, offset, local_track_name)
        track.operational_points.append(op_part)
        self.parts.append(op_part)


@dataclass
class OperationalPointPart:
    operational_point: OperationalPoint
    position: float
    local_track_name: str

    def to_rjs(self, track):
        return models.OperationalPointPart(
            track=track.id,
            position=self.position,
            local_track_name=self.local_track_name,
        )

File: editoast/schemas/src/infra/operational_point.rs
use crate::primitives::NonBlankString;
use educe::Educe;
use serde::Deserialize;
use serde::Serialize;
use utoipa::ToSchema;

use super::TrackOffset;
use crate::primitives::Identifier;
use crate::primitives::OSRDIdentified;
use crate::primitives::OSRDTyped;
use crate::primitives::ObjectType;

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct OperationalPoint {
    #[schema(inline)]
    pub id: Identifier,
    pub parts: Vec<OperationalPointPart>,
    #[serde(default)]
    pub weight: Option<u8>,
    #[schema(inline)]
    pub name: NonBlankString,
    pub uic: Option<u32>,
    /// Primary Location Code : https://rne.eu/it/products/ccs/crd/
    #[schema(inline)]
    pub plc: Option<NonBlankString>,
    #[schema(inline)]
    pub country_code: NonBlankString,
    #[schema(inline)]
    pub main_code: NonBlankString,
    #[schema(inline)]
    pub secondary_code: Option<NonBlankString>,
    pub is_passenger_station: bool,
    #[schema(inline)]
    pub secondary_name: Option<NonBlankString>,
}

#[derive(Debug, Educe, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
#[serde(deny_unknown_fields)]
#[educe(Default)]
pub struct OperationalPointPart {
    #[educe(Default = "InvalidRef".into())]
    #[schema(inline)]
    pub track: Identifier,
    /// Offset on the track section, in m
    pub position: f64,
    pub local_track_name: NonBlankString,
    #[serde(default)]
    pub extensions: OperationalPointPartExtension,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct OperationalPointPartExtension {
    pub sncf: Option<OperationalPointPartSncfExtension>,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct OperationalPointPartSncfExtension {
    pub kp: String,
}

impl OSRDTyped for OperationalPoint {
    fn get_type() -> ObjectType {
        ObjectType::OperationalPoint
    }
}

impl OSRDIdentified for OperationalPoint {
    fn get_id(&self) -> &String {
        &self.id
    }
}

impl OperationalPoint {
    pub fn track_offsets(&self) -> Vec<TrackOffset> {
        self.track_offsets_by_local_track_name(None)
    }

    pub fn track_offsets_by_local_track_name(
        &self,
        local_track_name: Option<&NonBlankString>,
    ) -> Vec<TrackOffset> {
        self.parts
            .iter()
            .filter(|part| {
                local_track_name
                    .is_none_or(|local_track_name| *local_track_name == part.local_track_name)
            })
            .map(|part| TrackOffset {
                track: part.track.clone(),
                offset: (part.position * 1000.0).round() as u64,
            })
            .collect()
    }
}
